File: Noxa.Emulation.Psp.Bios.FastHLE/Modules/ThreadManForUser_ThreadControl.cpp
// ----------------------------------------------------------------------------
// PSP Player Emulation Suite
// Copyright (C) 2006 Ben Vanik (noxa)
// Licensed under the LGPL - see License.txt in the project root for details
// ----------------------------------------------------------------------------

#include "Stdafx.h"
#include "ThreadManForUser.h"
#include "Kernel.h"
#include "KernelHelpers.h"
#include "KernelPartition.h"
#include "KernelStatistics.h"
#include "KernelThread.h"

using namespace System;
using namespace System::Diagnostics;
using namespace Noxa::Emulation::Psp;
using namespace Noxa::Emulation::Psp::Bios;
using namespace Noxa::Emulation::Psp::Bios::Modules;

// int sceKernelSleepThread(); (/user/pspthreadman.h:244)
int ThreadManForUser::sceKernelSleepThread()
{
	KernelThread^ thread = _kernel->ActiveThread;
	if( thread == nullptr )
		return -1;

	thread->State = KernelThreadState::Waiting;
	thread->WaitID = 0;
	thread->WaitClass = KernelThreadWait::Sleep;
	thread->CanHandleCallbacks = false;
	_kernel->ContextSwitch();

	return 0;
}

// int sceKernelSleepThreadCB(); (/user/pspthreadman.h:255)
int ThreadManForUser::sceKernelSleepThreadCB()
{
	KernelThread^ thread = _kernel->ActiveThread;
	if( thread == nullptr )
		return -1;

	thread->State = KernelThreadState::Waiting;
	thread->WaitID = 0;
	thread->WaitClass = KernelThreadWait::Sleep;
	thread->CanHandleCallbacks = true;
	_kernel->ContextSwitch();

	return 0;
}

// int sceKernelWakeupThread(SceUID thid); (/user/pspthreadman.h:264)
int ThreadManForUser::sceKernelWakeupThread( int thid )
{
	KernelThread^ thread = _kernel->FindThread( thid );
	if( thread == nullptr )
		return -1;

	thread->State = KernelThreadState::Ready;
	thread->WakeupCount++;

	_kernel->ContextSwitch();

	return 0;
}

// int sceKernelCancelWakeupThread(SceUID thid); (/user/pspthreadman.h:273)
int ThreadManForUser::sceKernelCancelWakeupThread( int thid )
{
	KernelThread^ thread = _kernel->FindThread( thid );
	if( thread == nullptr )
		return -1;

	// cancel wakeup not supported - perhaps we shouldn't cs in sceKernelWakeupThread?
	Debug::Assert( false );

	if( thread->State == KernelThreadState::Ready )
	{
		thread->State = KernelThreadState::Waiting;
		thread->WaitID = 0;
		thread->WaitClass = KernelThreadWait::Sleep;

		return 0;
	}
	else
		return -1;
}

// int sceKernelSuspendThread(SceUID thid); (/user/pspthreadman.h:282)
int ThreadManForUser::sceKernelSuspendThread( int thid )
{
	KernelThread^ thread = _kernel->FindThread( thid );
	if( thread == nullptr )
		return -1;

	thread->State = KernelThreadState::Suspended;

	_kernel->ContextSwitch();

	return 0;
}

// int sceKernelResumeThread(SceUID thid); (/user/pspthreadman.h:291)
int ThreadManForUser::sceKernelResumeThread( int thid )
{
	KernelThread^ thread = _kernel->FindThread( thid );
	if( thread == nullptr )
		return -1;

	thread->State = KernelThreadState::Ready;

	_kernel->ContextSwitch();

	return 0;
}

// int sceKernelWaitThreadEnd(SceUID thid, SceUInt *timeout); (/user/pspthreadman.h:301)
int ThreadManForUser::sceKernelWaitThreadEnd( IMemory^ memory, int thid, int timeout )
{
	KernelThread^ thread = _kernel->FindThread( thid );
	if( thread == nullptr )
		return -1;

	thread->State = KernelThreadState::Waiting;
	thread->WaitID = thid;
	thread->WaitClass = KernelThreadWait::ThreadEnd;
	if( timeout != 0x0 )
		thread->WaitTimeout = memory->ReadWord( timeout );
	else
		thread->WaitTimeout = 0;
	thread->CanHandleCallbacks = false;

	_kernel->ContextSwitch();

	return 0;
}

// int sceKernelWaitThreadEndCB(SceUID thid, SceUInt *timeout); (/user/pspthreadman.h:311)
int ThreadManForUser::sceKernelWaitThreadEndCB( IMemory^ memory, int thid, int timeout )
{
	KernelThread^ thread = _kernel->FindThread( thid );
	if( thread == nullptr )
		return -1;

	thread->State = KernelThreadState::Waiting;
	thread->WaitID = thid;
	thread->WaitClass = KernelThreadWait::ThreadEnd;
	if( timeout != 0x0 )
		thread->WaitTimeout = memory->ReadWord( timeout );
	else
		thread->WaitTimeout = 0;
	thread->CanHandleCallbacks = true;

	_kernel->ContextSwitch();

	return 0;
}

int ThreadManForUser::ThreadDelayComparer( KernelThread^ a, KernelThread^ b )
{
	int64 aend = a->WaitTimestamp + a->WaitTimeout;
	int64 bend = b->WaitTimestamp + a->WaitTimeout;
	return aend.CompareTo( bend );
}

// int sceKernelDelayThread(SceUInt delay); (/user/pspthreadman.h:323)
int ThreadManForUser::sceKernelDelayThread( int delay )
{
	KernelThread^ thread = _kernel->ActiveThread;
	if( thread == nullptr )
		return -1;

	thread->State = KernelThreadState::Waiting;
	thread->WaitClass = KernelThreadWait::Delay;
	thread->WaitID = 0;
	thread->WaitTimeout = delay * 10; // in us - convert to ticks
	thread->WaitTimestamp = DateTime::Now.Ticks;
	thread->CanHandleCallbacks = false;

	_kernel->_delayedThreads->Add( thread );
	_kernel->_delayedThreads->Sort( gcnew Comparison<KernelThread^>( this, &ThreadManForUser::ThreadDelayComparer ) );

	if( _kernel->_delayedThreadTimer->Enabled == false )
		_kernel->SpawnDelayedThreadTimer( thread->WaitTimeout + thread->WaitTimestamp );

	_kernel->ContextSwitch();
	
	return 0;
}

// int sceKernelDelayThreadCB(SceUInt delay); (/user/pspthreadman.h:335)
int ThreadManForUser::sceKernelDelayThreadCB( int delay )
{
	KernelThread^ thread = _kernel->ActiveThread;
	if( thread == nullptr )
		return -1;

	thread->State = KernelThreadState::Waiting;
	thread->WaitClass = KernelThreadWait::Delay;
	thread->WaitID = 0;
	thread->WaitTimeout = delay * 10; // us->ticks
	thread->WaitTimestamp = DateTime::Now.Ticks;
	thread->CanHandleCallbacks = true;

	_kernel->_delayedThreads->Add( thread );
	_kernel->_delayedThreads->Sort( gcnew Comparison<KernelThread^>( this, &ThreadManForUser::ThreadDelayComparer ) );

	if( _kernel->_delayedThreadTimer->Enabled == false )
		_kernel->SpawnDelayedThreadTimer( thread->WaitTimeout + thread->WaitTimestamp );

	_kernel->ContextSwitch();
	
	return 0;
}

// int sceKernelDelaySysClockThread(SceKernelSysClock *delay); (/user/pspthreadman.h:344)
int ThreadManForUser::sceKernelDelaySysClockThread( IMemory^ memory, int delay )
{
	KernelThread^ thread = _kernel->ActiveThread;
	if( thread == nullptr )
		return -1;
	
	return NISTUBRETURN;
}

// int sceKernelDelaySysClockThreadCB(SceKernelSysClock *delay); (/user/pspthreadman.h:354)
int ThreadManForUser::sceKernelDelaySysClockThreadCB( IMemory^ memory, int delay )
{
	KernelThread^ thread = _kernel->ActiveThread;
	if( thread == nullptr )
		return -1;
	
	return NISTUBRETURN;
}

File: trunk/Noxa.Emulation.Psp.Bios.ManagedHLE/Modules/ThreadManForUser.ThreadControl.cs
// ----------------------------------------------------------------------------
// PSP Player Emulation Suite
// Copyright (C) 2006 Ben Vanik (noxa)
// Licensed under the LGPL - see License.txt in the project root for details
// ----------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Text;

using Noxa.Utilities;
using Noxa.Emulation.Psp;
using Noxa.Emulation.Psp.Bios;
using Noxa.Emulation.Psp.Cpu;

namespace Noxa.Emulation.Psp.Bios.ManagedHLE.Modules
{
	partial class ThreadManForUser
	{
		[NotImplemented]
		[Stateless]
		[BiosFunction( 0x9ACE131E, "sceKernelSleepThread" )]
		// SDK location: /user/pspthreadman.h:244
		// SDK declaration: int sceKernelSleepThread();
		public int sceKernelSleepThread()
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0x82826F70, "sceKernelSleepThreadCB" )]
		// SDK location: /user/pspthreadman.h:255
		// SDK declaration: int sceKernelSleepThreadCB();
		public int sceKernelSleepThreadCB()
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0xD59EAD2F, "sceKernelWakeupThread" )]
		// SDK location: /user/pspthreadman.h:264
		// SDK declaration: int sceKernelWakeupThread(SceUID thid);
		public int sceKernelWakeupThread( int thid )
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0xFCCFAD26, "sceKernelCancelWakeupThread" )]
		// SDK location: /user/pspthreadman.h:273
		// SDK declaration: int sceKernelCancelWakeupThread(SceUID thid);
		public int sceKernelCancelWakeupThread( int thid )
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0x9944F31F, "sceKernelSuspendThread" )]
		// SDK location: /user/pspthreadman.h:282
		// SDK declaration: int sceKernelSuspendThread(SceUID thid);
		public int sceKernelSuspendThread( int thid )
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0x75156E8F, "sceKernelResumeThread" )]
		// SDK location: /user/pspthreadman.h:291
		// SDK declaration: int sceKernelResumeThread(SceUID thid);
		public int sceKernelResumeThread( int thid )
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0x278C0DF5, "sceKernelWaitThreadEnd" )]
		// SDK location: /user/pspthreadman.h:301
		// SDK declaration: int sceKernelWaitThreadEnd(SceUID thid, SceUInt *timeout);
		public int sceKernelWaitThreadEnd( int thid, int timeout )
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0x840E8133, "sceKernelWaitThreadEndCB" )]
		// SDK location: /user/pspthreadman.h:311
		// SDK declaration: int sceKernelWaitThreadEndCB(SceUID thid, SceUInt *timeout);
		public int sceKernelWaitThreadEndCB( int thid, int timeout )
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0xCEADEB47, "sceKernelDelayThread" )]
		// SDK location: /user/pspthreadman.h:323
		// SDK declaration: int sceKernelDelayThread(SceUInt delay);
		public int sceKernelDelayThread( int delay )
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0x68DA9E36, "sceKernelDelayThreadCB" )]
		// SDK location: /user/pspthreadman.h:335
		// SDK declaration: int sceKernelDelayThreadCB(SceUInt delay);
		public int sceKernelDelayThreadCB( int delay )
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0xBD123D9E, "sceKernelDelaySysClockThread" )]
		// SDK location: /user/pspthreadman.h:344
		// SDK declaration: int sceKernelDelaySysClockThread(SceKernelSysClock *delay);
		public int sceKernelDelaySysClockThread( int delay )
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0x1181E963, "sceKernelDelaySysClockThreadCB" )]
		// SDK location: /user/pspthreadman.h:354
		// SDK declaration: int sceKernelDelaySysClockThreadCB(SceKernelSysClock *delay);
		public int sceKernelDelaySysClockThreadCB( int delay )
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0x2C34E053, "sceKernelReleaseWaitThread" )]
		// SDK location: /user/pspthreadman.h:399
		// SDK declaration: int sceKernelReleaseWaitThread(SceUID thid);
		public int sceKernelReleaseWaitThread( int thid )
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0x3AD58B8C, "sceKernelSuspendDispatchThread" )]
		// SDK location: /user/pspthreadman.h:227
		// SDK declaration: int sceKernelSuspendDispatchThread();
		public int sceKernelSuspendDispatchThread()
		{
			return Module.NotImplementedReturn;
		}

		[NotImplemented]
		[Stateless]
		[BiosFunction( 0x27E22EC2, "sceKernelResumeDispatchThread" )]
		// SDK location: /user/pspthreadman.h:237
		// SDK declaration: int sceKernelResumeDispatchThread(int state);
		public int sceKernelResumeDispatchThread( int state )
		{
			return Module.NotImplementedReturn;
		}
	}
}
